### PR TITLE
Fluentd: Compress fluentd.conf file

### PR DESCRIPTION
**Pull request by zvlb**
Signed-off-by: Zemtsov Vladimir <vl.zemtsov@gmail.com>

| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | no
| License         | Apache 2.0

### What's in this PR?
Use new config-reloader and compress fluentd.conf (gzip)

### Why?
In my K8s cluster I have many Flows (more then 2000) and Logging-operator can't create secret for fluentd.conf, because it has a weight of more than 3 MB. (https://kubernetes.io/docs/concepts/configuration/secret/#restriction-data-size)

I create new application for config-reloader - https://github.com/zvlb/config-reloader. it can unzip files in secret and move file to correct directory (like it's do in [prometheus-config-reloader](https://github.com/prometheus-operator/prometheus-operator/pkgs/container/prometheus-config-reloader))

In this PR I update:
- Add bool field `compressConfigFile` to CRD Logging (spec.fluentd.compressConfigFile)
- Change config-reloader image from `jimmidyson/configmap-reload` to `zvlb:config-reloader`
- If  `spec.fluentd.compressConfigFile` - true:
    - Logging operator [compress](https://github.com/zvlb/logging-operator/blob/gzip-fluentd-conf/pkg/resources/fluentd/appconfigmap.go#L41) Fluentd.conf before save it in Kubernetes Secret
    - For fluentd-configcheck created initContainer, where fluentd.conf will be unarchived
    - For fluentd created initContainer, where fluentd.conf will be unarchived, and update config-reloader sidecar

### Additional context
I run this fork on my K8s cluster, where more then 2000 flows. All work correctly.
Weight fluentd.conf - 6M
Weight fluentd.conf.gz (in secret) - 528K

### Checklist

- [x] Code meets the [Developer Guide](https://github.com/banzaicloud/developer-guide)
- [x] User guide and development docs updated (if needed)
- [x] Related Helm chart(s) updated (if needed)

### CODE REVIEW

**Comment by zvlb** [comment]
In test `CI / build (pull_request)` check different in CRD `Loggings`
I update this CRD and can't pass checks

**Comment by zvlb** [comment]
Oh.... Its works 1 day, and after I have error:
```
"msg":"Reconciler error","reconciler group":"logging.banzaicloud.io","reconciler kind":"Logging","name":"logging-operator-logging","namespace":"","error":"failed to reconcile resource: failed to create resource: creating resource failed: Secret \"logging-operator-logging-fluentd-app\" is invalid: metadata.annotations: Too long: must have at most 262144 bytes"
```

It's happens, bc logging operator create `banzaicloud.com/last-applied` annotation with fluentd.conf data (zipped, but more then 262144 bytes

I don't know what I can do with this....

**Comment by zvlb** [comment]
> Oh.... Its works 1 day, and after I have error:
> 
> ```
> "msg":"Reconciler error","reconciler group":"logging.banzaicloud.io","reconciler kind":"Logging","name":"logging-operator-logging","namespace":"","error":"failed to reconcile resource: failed to create resource: creating resource failed: Secret \"logging-operator-logging-fluentd-app\" is invalid: metadata.annotations: Too long: must have at most 262144 bytes"
> ```
> 
> It's happens, bc logging operator create `banzaicloud.com/last-applied` annotation with fluentd.conf data (zipped, but more then 262144 bytes
> 
> I don't know what I can do with this....

Ok... I [add](https://github.com/zvlb/logging-operator/commit/ed552ddce6d05d9e2295d615f65b9445abe36a3a) custom reconciler for secret with generated fluentd.conf. I don't add this in PR. But if you want - I can do it

**Comment by LinkMaq** [comment]
Yes, I had a similar problem after creating only 50 flows in my cluster.There is also a similar problem
[banzaicloud.com/last-applied too long on FluentD configuration secret](https://github.com/banzaicloud/logging-operator/issues/837) #837 

This solution is great!  And it is the same idea as proemtheus-operator，I can't wait to use this feature! 😄

**Comment by zvlb** [comment]
Hi. Can we finish this PR?
Needs to add my commit with custom reconcile for secret with fluentd config?

File: pkg/sdk/logging/api/v1beta1/logging_types.go [hunk 1]
```diff
@@ -131,8 +131,8 @@ const (
 	DefaultFluentdDrainPauseImageTag            = "latest"
 	DefaultFluentdVolumeModeImageRepository     = "busybox"
 	DefaultFluentdVolumeModeImageTag            = "latest"
-	DefaultFluentdConfigReloaderImageRepository = "jimmidyson/configmap-reload"
-	DefaultFluentdConfigReloaderImageTag        = "v0.4.0"
+	DefaultFluentdConfigReloaderImageRepository = "zvlb/config-reloader"
```
Comment (ahma) [inline pkg/sdk/logging/api/v1beta1/logging_types.go:134]:
```suggestion
	DefaultFluentdConfigReloaderImageRepository = "ghcr.io/banzaicloud/config-reloader"
```

File: pkg/sdk/logging/api/v1beta1/logging_types.go [hunk 2]
```diff
@@ -131,8 +131,8 @@ const (
...
+	DefaultFluentdConfigReloaderImageTag        = "v0.1.0"
```
Comment (ahma) [inline pkg/sdk/logging/api/v1beta1/logging_types.go:135]:
```suggestion
	DefaultFluentdConfigReloaderImageTag        = "0.0.1"
```

**Comment by ahma** [comment]
Thanks for the hard work @zvlb, and sorry for the late reply. 
The only thing I would like to ask is to please change the source of the image. I hope this is understandable. If you have any questions because of this please let me know.

**Comment by zvlb** [comment]
I change the source for config-reloader to our fork - no problem.
Maybe I migrate config-reloader to the corporate account, and use fork - it's a good idea!

I have some ideas on how to do config-reloader better. I will notify you if I do this.

**Comment by zvlb** [comment]
And what do you think about [this](https://github.com/zvlb/logging-operator/commit/ed552ddce6d05d9e2295d615f65b9445abe36a3a) solution for reconciling app-config secret?

We discuss this earlier. 
Or I can open new PR for this problem

**Review by siliconbrain — APPROVED** [review]
Thank you for implementing this!

> I have some ideas on how to do config-reloader better. I will notify you if I do this.

That would be great! Feel free to open PRs under [banzaicloud/config-reloader](github.com/banzaicloud/config-reloader).

> And what do you think about [this](https://github.com/zvlb/logging-operator/commit/ed552ddce6d05d9e2295d615f65b9445abe36a3a) solution for reconciling app-config secret?
> 
> We discuss this earlier. 
> Or I can open new PR for this problem

I would like to try something first to get rid of the root of the problem with `banzaicloud.com/last-applied` altogether, but if that fails we'd appreciate the PR. If you want, you can open the PR and we'll either close it or merge it depending on which solution works out.

**Comment by zvlb** [comment]
> I would like to try something first to get rid of the root of the problem with `banzaicloud.com/last-applied` altogether, but if that fails we'd appreciate the PR. If you want, you can open the PR and we'll either close it or merge it depending on which solution works out.

I have some solution for this problem.
For example, you can use hash/crc32 to compare hash. (Not clean configs in []byte). (How we do this for [Vector Operator](https://github.com/kaasops/vector-operator/blob/main/controllers/vectorpipeline_controller.go#L74) 

I open a new PR for this issue. (Maybe next week)

**Comment by ahma** [comment]
@zvlb Could you please run a `make generate` on your branch, please?

**Review by ahma — APPROVED** [review]
LGTM thanks @zvlb

**Review by siliconbrain — APPROVED** [review]

**Pull request merged**